File: jtaf-web/src/main/java/ch/jtaf/report/Numbers.java
package ch.jtaf.report;

import ch.jtaf.entity.Athlete;
import com.itextpdf.text.Document;
import com.itextpdf.text.DocumentException;
import com.itextpdf.text.FontFactory;
import com.itextpdf.text.PageSize;
import com.itextpdf.text.Phrase;
import com.itextpdf.text.pdf.PdfPCell;
import com.itextpdf.text.pdf.PdfPTable;
import com.itextpdf.text.pdf.PdfWriter;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.List;
import java.util.Locale;

public class Numbers extends AbstractReport {

    private final static float FONT_SIZE_INFO = 12f;
    private final static float FONT_SIZE_TEXT = 90f;
    private final List<Athlete> athletes;

    public Numbers(List<Athlete> athletes, Locale locale) {
        super(locale);
        this.athletes = athletes;
    }

    public byte[] create() {
        try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
            Document document = new Document(PageSize.A4, cmToPixel(1.5f), cmToPixel(1.7f), cmToPixel(0.8f), cmToPixel(0f));
            PdfWriter pdfWriter = PdfWriter.getInstance(document, baos);
            document.open();
            int i = 0;
            int number = 1;
            PdfPTable table = createMainTable();
            for (Athlete athlete : athletes) {
                if (i > 9) {
                    document.add(table);
                    document.newPage();
                    i = 0;
                    table = createMainTable();
                }
                addAthleteInfo(table, athlete, number);
                if (i % 2 == 0) {
                    addEmptyCell(table);
                }
                if (i == 1 || i == 3 || i == 5 || i == 7) {
                    addEmptyRow(table);
                }
                i++;
                number++;
            }
            document.add(table);
            document.close();
            pdfWriter.flush();
            return baos.toByteArray();
        } catch (DocumentException | IOException e) {
            e.printStackTrace();
            return new byte[0];
        }
    }

    private PdfPTable createMainTable() {
        PdfPTable table = new PdfPTable(new float[]{10f, 1.8f, 10f});
        table.setWidthPercentage(100);
        return table;
    }

    private void addEmptyCell(PdfPTable table) {
        PdfPCell cellEmpty = new PdfPCell(new Phrase(""));
        cellEmpty.setBorder(0);
        table.addCell(cellEmpty);
    }

    private void addAthleteInfo(PdfPTable table, Athlete athlete, int number) {
        PdfPTable atable = new PdfPTable(1);
        atable.setWidthPercentage(100);

        PdfPCell cellId = new PdfPCell(
                new Phrase(number + "",
                        FontFactory.getFont(FontFactory.HELVETICA, FONT_SIZE_TEXT)));
        cellId.setBorder(0);
        cellId.setMinimumHeight(cmToPixel(2.5f));
        cellId.setHorizontalAlignment(PdfPCell.ALIGN_CENTER);
        atable.addCell(cellId);

        String text = athlete.getLastName() + " " + athlete.getFirstName() + "\n";
        text += athlete.getCategory().getAbbreviation();
        if (athlete.getClub() != null) {
            text += " / " + athlete.getClub().getName();
        }

        PdfPCell cellName = new PdfPCell(new Phrase(text, FontFactory.getFont(FontFactory.HELVETICA, FONT_SIZE_INFO)));
        cellName.setBorder(0);
        cellName.setMinimumHeight(cmToPixel(1.8f));
        cellName.setHorizontalAlignment(PdfPCell.ALIGN_CENTER);
        cellName.setVerticalAlignment(PdfPCell.ALIGN_MIDDLE);
        atable.addCell(cellName);

        PdfPCell cellTable = new PdfPCell(atable);
        cellTable.setBorder(0);

        table.addCell(cellTable);
    }

    private void addEmptyRow(PdfPTable table) {
        PdfPCell cellId = new PdfPCell(new Phrase(" "));
        cellId.setBorder(0);
        cellId.setMinimumHeight(cmToPixel(0.5f));
        cellId.setHorizontalAlignment(PdfPCell.ALIGN_CENTER);
        table.addCell(cellId);

        addEmptyCell(table);

        PdfPCell cellName = new PdfPCell(new Phrase(" "));
        cellName.setBorder(0);
        cellName.setMinimumHeight(cmToPixel(0.5f));
        cellName.setHorizontalAlignment(PdfPCell.ALIGN_CENTER);
        table.addCell(cellName);
    }

}

File: src/main/kotlin/ch/jtaf/control/reporting/report/Numbers.kt
package ch.jtaf.control.reporting.report

import ch.jtaf.entity.AthleteDTO
import com.itextpdf.text.Document
import com.itextpdf.text.FontFactory
import com.itextpdf.text.PageSize
import com.itextpdf.text.Phrase
import com.itextpdf.text.pdf.PdfPCell
import com.itextpdf.text.pdf.PdfPTable
import com.itextpdf.text.pdf.PdfWriter
import java.io.ByteArrayOutputStream
import java.util.*

class Numbers(private val athletes: List<AthleteDTO>, locale: Locale) : AbstractReport(locale) {

    private val fontSizeInfo = 12f
    private val fontSizeText = 90f

    fun create(): ByteArray {
        try {
            ByteArrayOutputStream().use { baos ->
                val document = Document(PageSize.A4, cmToPixel(1.5f), cmToPixel(1.7f), cmToPixel(0.8f), cmToPixel(0f))
                val pdfWriter = PdfWriter.getInstance(document, baos)
                document.open()
                var i = 0
                var number = 1
                var table = createMainTable()
                athletes.forEach {
                    if (i > 9) {
                        document.add(table)
                        document.newPage()
                        i = 0
                        table = createMainTable()
                    }
                    addAthleteInfo(table, it, number)
                    if (i % 2 == 0) {
                        addEmptyCell(table)
                    }
                    if (i == 1 || i == 3 || i == 5 || i == 7) {
                        addEmptyRow(table)
                    }
                    i++
                    number++
                }
                document.add(table)
                document.close()
                pdfWriter.flush()
                return baos.toByteArray()
            }
        } catch (e: Exception) {
            e.printStackTrace()
            return ByteArray(0)
        }

    }

    private fun createMainTable(): PdfPTable {
        val table = PdfPTable(floatArrayOf(10f, 1.8f, 10f))
        table.widthPercentage = 100f
        return table
    }

    private fun addEmptyCell(table: PdfPTable) {
        val cellEmpty = PdfPCell(Phrase(""))
        cellEmpty.border = 0
        table.addCell(cellEmpty)
    }

    private fun addAthleteInfo(table: PdfPTable, athlete: AthleteDTO, number: Int) {
        val atable = PdfPTable(1)
        atable.widthPercentage = 100f

        val cellId = PdfPCell(Phrase(number.toString() + "", FontFactory.getFont(FontFactory.HELVETICA, fontSizeText)))
        cellId.border = 0
        cellId.minimumHeight = cmToPixel(2.5f)
        cellId.horizontalAlignment = PdfPCell.ALIGN_CENTER
        atable.addCell(cellId)

        var text = athlete.lastName + " " + athlete.firstName + "\n"
        text += athlete.category
        if (athlete.club != null) {
            text += " / " + athlete.club
        }

        val cellName = PdfPCell(Phrase(text, FontFactory.getFont(FontFactory.HELVETICA, fontSizeInfo)))
        cellName.border = 0
        cellName.minimumHeight = cmToPixel(1.8f)
        cellName.horizontalAlignment = PdfPCell.ALIGN_CENTER
        cellName.verticalAlignment = PdfPCell.ALIGN_MIDDLE
        atable.addCell(cellName)

        val cellTable = PdfPCell(atable)
        cellTable.border = 0

        table.addCell(cellTable)
    }

    private fun addEmptyRow(table: PdfPTable) {
        val cellId = PdfPCell(Phrase(" "))
        cellId.border = 0
        cellId.minimumHeight = cmToPixel(0.5f)
        cellId.horizontalAlignment = PdfPCell.ALIGN_CENTER
        table.addCell(cellId)

        addEmptyCell(table)

        val cellName = PdfPCell(Phrase(" "))
        cellName.border = 0
        cellName.minimumHeight = cmToPixel(0.5f)
        cellName.horizontalAlignment = PdfPCell.ALIGN_CENTER
        table.addCell(cellName)
    }

}
